candle-core/src/tensor.rs

```diff
@@ -1235,6 +1235,74 @@ impl Tensor {
         Ok(from_storage(storage, (n, c, h_out, w_out), op, false))
     }
 
+    /// Computes the dot product between two tensors
+    ///
+    /// - If inputs are 1D vectors (`[n]`), returns their scalar dot product.
+    /// - If inputs are 2D matrices (`[m, n]` and `[m, n]`), returns the Frobenius inner product
+    ///   (sum of element-wise products).
+    /// - Panics if shapes are incompatible.
+    ///
+    /// # Example (vectors)
+    /// ```rust
+    /// let a = Tensor::new(&[1.0, 2.0, 3.0], &Device::Cpu)?;
+    /// let b = Tensor::new(&[4.0, 5.0, 6.0], &Device::Cpu)?;
+    /// let res = a.dot(&b)?;  // 1 * 4 + 2 * 5 + 3 * 6 = 32
+    /// ```
+    pub fn dot(&self, rhs: &Self) -> Result<Self> {
+        if self.shape() != rhs.shape() {
+            return Err(Error::ShapeMismatchBinaryOp {
+                lhs: self.shape().clone(),
+                rhs: rhs.shape().clone(),
+                op: "dot",
+            });
+        }
+        (self * rhs)
+            .and_then(|ret| ret.sum_all())
+    }
+
+    /// Computes the **Frobenius norm** (L2 norm of all elements) of the tensor.
```

Comment:
Returning an int result on an int tensor seems pretty error prone as the output will be rounded + it's easy to do computations that would result in overflowing on u8 or even u32. I would suggest also being in line with the pytorch behavior here so bail on integer type tensors.